src/equihash.rs

```diff
@@ -0,0 +1,471 @@
+use blake2_rfc::blake2b::{Blake2b, Blake2bResult};
+use byteorder::{BigEndian, LittleEndian, ReadBytesExt, WriteBytesExt};
+use std::io::Cursor;
+use std::mem::size_of;
+
+struct Params {
+    n: u32,
+    k: u32,
+}
+
+#[derive(Clone)]
+struct Node {
+    hash: Vec<u8>,
+    indices: Vec<u32>,
+}
+
+impl Params {
+    fn indices_per_hash_output(&self) -> u32 {
+        512 / self.n
+    }
+    fn hash_output(&self) -> u8 {
+        (self.indices_per_hash_output() * self.n / 8) as u8
+    }
+    fn collision_bit_length(&self) -> usize {
+        (self.n / (self.k + 1)) as usize
+    }
+    fn collision_byte_length(&self) -> usize {
+        (self.collision_bit_length() + 7) / 8
+    }
+    fn hash_length(&self) -> usize {
+        ((self.k as usize) + 1) * self.collision_byte_length()
+    }
+}
+
+impl Node {
+    fn from_children(a: Node, b: Node, trim: usize) -> Self {
+        let hash: Vec<_> = a.hash
+            .iter()
+            .zip(b.hash.iter())
+            .skip(trim)
+            .map(|(a, b)| a ^ b)
+            .collect();
+        let indices = if a.indices_before(&b) {
+            let mut indices = a.indices;
```

Comment:
I think that on average this will end up performing a re-allocation. I wouldn't be surprised if it's just as efficient to always use `from_children_ref`.